### PR TITLE
Add support for GraalVM native serial GC logs

### PR DESCRIPTION
**Pull request by galderz**
This PR enhances the repository to read the serial GC logs in GraalVM native runtime, SubstrateVM.

It adds:
* Support for reading GraalVM serial GC logs.
* Unit test to verify that the elements are retrieved as expected.

GraalVM's serial GC is stop-the-world, so both the `Incremental GC` and `Full GC` should be considered pauses. To evaluate the data in the `GraalVM.log`, the `pauses_only` method needs to look like this:

```
    def pauses_only(row):
      if "EventType" in row:
          if row["EventType"] == "Pause":
              return True

      if "EventName" in row:
          if row["EventName"] == "Incremental GC" or row["EventName"] == "Full GC":
              return True

    return False
```

Having done that, and analysing the `GraalVM.log` file independently, you get graphs like these:

STW pauses during program time
STW pauses by Event Name
Heap used before GC
Heap used after GC

The PR does not currently update the notebook because the workload for the sample `GraalVM.log` is not comparable to the workload of the other GC logs. I want to run some tests where I do such a thing and I can compare all, including GraalVM, GC logs. This probably a new dataset, say stored under `datasets/quarkus` and maybe a different notebook to not overwrite existing notebook. Any thoughts on this?

### CODE REVIEW

**Comment by ivankrylov** [comment]
Just a quick note that I noticed the PR and I will get back shortly.

File: src/parse_log_file.py [hunk 1]
```diff
@@ -22,12 +22,20 @@ def get_parsing_groups():
...
+
+    colon = ": ", None, None
+
+    space = "\s", None, None
+
+    date_time = "(\d{4}-\d\d-\d\dT\d\d:\d\d:\d\d\.\d{3}[+-]\d{4})", "DateTime", str # [9999-08-26T14:42:00.565-0400]
                                                                                         # [2021-08-26T14:42:59.565-0400]
-    time = "\[([\d\.]+)", "Time", float # 999999
+    time = "([\d\.]+)", "Time", float # 999999
                                         # 123541.21425
```
Comment (bmathiske) [inline src/parse_log_file.py:37]:
```suggestion
                                      # 123541.21425
```

File: src/parse_log_file.py [hunk 2]
```diff
@@ -60,6 +68,15 @@ def get_parsing_groups():
                                                             # ->99999M
     zgc_max_heapsize = "\((\d+)%\)", "HeapPercentFull", float   # (24%)
                                                             # (00000%)
+
+    graalvm_heap_before_gc = "(\d+)K", "HeapBeforeGC_kb", float # 123K
+                                                                       # 9999K
```
Comment (bmathiske) [inline src/parse_log_file.py:73]:
```suggestion
                                                                # 9999K
```

File: src/parse_log_file.py [hunk 3]
```diff
@@ -60,6 +68,15 @@ def get_parsing_groups():
...
+    graalvm_heap_after_gc = "->(\d+)K, ", "HeapAfterGC_kb", float # ->200K
+                                                                         # ->9999K
```
Comment (bmathiske) [inline src/parse_log_file.py:75]:
```suggestion
                                                                  # ->9999K
```

File: src/parse_log_file.py [hunk 4]
```diff
@@ -60,6 +68,15 @@ def get_parsing_groups():
...
+
+    graalvm_duration_sec = "([\d.]+) secs", "Duration_seconds", float # 99999.9999 secs
+                                                                     # 0.0 secs
```
Comment (bmathiske) [inline src/parse_log_file.py:78]:
```suggestion
                                                                      # 0.0 secs
```

File: src/read_log_file.py [hunk 5]
```diff
@@ -96,9 +98,46 @@ def scale_time(df):
...
+    df = scale_heap_unit(df, "HeapAfterGC_kb", "HeapAfterGC")
+    return df
+
+
+def scale_heap_unit(df, from_column, to_column):
+    heap_mb = []
+    for row in df[from_column]:
+        if row != None:
+            heap_mb.append(row * 1024)
+
```
Comment (bmathiske) [inline src/read_log_file.py:133]:
```suggestion
```

File: src/read_log_file.py [hunk 6]
```diff
@@ -96,9 +98,46 @@ def scale_time(df):
...
+    if heap_mb:
+        df[to_column] = heap_mb
+        df = df.drop(columns=[from_column], axis=1)
+
```
Comment (bmathiske) [inline src/read_log_file.py:137]:
```suggestion
```

File: src/read_log_file.py [hunk 7]
```diff
@@ -96,9 +98,46 @@ def scale_time(df):
...
     df["TimeFromStart_seconds"] = time_seconds
     df = df.drop(columns=["Time", "TimeUnit"], axis = 1)
-    return df        
+    return df
+
+# Create a column "Duration_milliseconds" that scales the duration in seconds to milliseconds
+def scale_duration(df):
+    if df.empty:
+        return df
+
```
Comment (bmathiske) [inline src/read_log_file.py:107]:
```suggestion
```

File: src/read_log_file.py [hunk 8]
```diff
@@ -96,9 +98,46 @@ def scale_time(df):
...
+    duration_milliseconds = []
+    for row in df["Duration_seconds"]:
+        if row != None:
+            duration_milliseconds.append(row * 1000)
+
```
Comment (bmathiske) [inline src/read_log_file.py:112]:
```suggestion
```

File: src/read_log_file.py [hunk 9]
```diff
@@ -96,9 +98,46 @@ def scale_time(df):
...
+    if duration_milliseconds:
+        df["Duration_milliseconds"] = duration_milliseconds
+        df = df.drop(columns=["Duration_seconds"], axis = 1)
+
```
Comment (bmathiske) [inline src/read_log_file.py:116]:
```suggestion
```

File: src/read_log_file.py (hunk 7 above)
Comment (bmathiske) [reply src/read_log_file.py:107]: Some methods have separating blank lines and another doesn't. I suggest using one of these styles uniformly. To better identify where a method starts and ends I am going without blank lines here. (If needed, in larger methods than these, one could insert almost blank lines with an empty '#' comment.)

File: src/read_log_file.py [hunk 10]
```diff
@@ -96,9 +98,46 @@ def scale_time(df):
...
+    return df
+
+# Create columns "HeapBeforeGC" and "HeapAfterGC" that scales the duration in kylobytes to megabytes
+def scale_heap_units(df):
```
Comment (bmathiske) [inline src/read_log_file.py:120]: Observing no blank lines here - see above.

**Comment by bmathiske** [comment]
I suggest changing method `pauses_only` in "GCGC.ipynb" to something like this:
```
def pauses_only(row):
    if "EventType" in row:
        return row["EventType"] == "Pause" or row["EventName"] == "Incremental GC" or row["EventName"] == "Full GC"
    return False 
```
Then users only need to change the input files and labels and have the existing notebook running for them in no time.

**Review by bmathiske — CHANGES_REQUESTED** [review]

**Comment by ivankrylov** [comment]
Hi @galderz , thank you so much for the contribution. 
As for the file: datasets/graalvm_log_file_test.log - can you kindly specify what version of GraalVM is used? Perhaps you could factor this file into a separate commit and specify in a commit message how this snipped is produced.

**Comment by galderz** [comment]
> As for the file: datasets/graalvm_log_file_test.log - can you kindly specify what version of GraalVM is used? Perhaps you could factor this file into a separate commit and specify in a commit message how this snipped is produced.

Fair enough. I'll look into it. I'll try to do the same with the existing `GraalVM.log` file, even though this one might change in the future as noted in the description.

**Comment by ivankrylov** [comment]
I am back from a somewhat long vacation. Let's drive this PR to integration. Can you factor the same vm go from GraalVm into a separate commit with a commit message that tells how exactly a log was produced? You can use any java app freely available to generate a log: a pet clinic app or simple app from a demo directory in Java 8 distributions.

**Comment by ivankrylov** [comment]
All right, I will take the change as is and address the remaining issues ourselves. 
@galderz Thank you so much for your contribution, keep them coming!

**Comment by galderz** [comment]
Hey @ivankrylov, thanks for integrating this!

Sorry I've not provided an update since last month. I've been working on some benchmarking and I was hoping to extract some GraalVM GC log out of it eventually for roughly a period of 5 minutes (that seems to be the period for the logs in the repo). However, some other things have come my way.

Also, the GC log output is changing in GraalVM for JDK 21 (to be released next week), so I'll send a separate PR to update the GC parsing logic to adjust for that version, and will use the opportunity to send a GC log that contains more interesting data.

**Comment by ivankrylov** [comment]
Sounds good @galderz! Looking forward for your next PR. This would also be a good chance to clear up in-the-sources notes on how GraalVM gc logs are generated (GraalVM version, test app name).

**Pull request merged**